### ISSUE TITLE
Multiple PowerShell tasks in a build/release causing timeouts with agent

### ISSUE THREAD

**Issue by begroff** [microsoft/azure-pipelines-tasks#14637]
## Note 
Issues in this repo are for tracking bugs, feature requests and questions for the tasks in this repo 

For a list:  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

If you have an issue or request for the Azure Pipelines service, use developer community instead:
https://developercommunity.visualstudio.com/spaces/21/index.html )

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: PowerShell / PowerShell on target machines

list here (V# not needed):   
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - TFS on-premises
    
    - If using TFS on-premises, provide the version: Dev18.M170.8

- Agent - Private: 

    - If using private agent, provide the OS of the machine running the agent and the agent version: 
        - OS: Windows 10 Pro 19042.867
        - Agent: 2.170.1

## Issue Description

My company recently updated Azure DevOps private server to the version above. We have both builds and releases that are using either the PowerShell (v2) task or the PowerShell on target machines (v3) task. When we have multiple tasks, the behavior we are experiencing is that the first PowerShell task runs fine, but when it gets to the second task it just sits there until we get the error message, **"We stopped hearing from agent xxxxx. Verify the agent machine is running and has a healthy network connection. Anything that terminates an agent process, starves it for CPU, or blocks its network access can cause this error."** We have restarted our build server and still continue to have this problem. It happens with either inline script or scripts in a file.

If we only have 1 PowerShell task in a build/release the problem does not happen. The tasks run fine.

![powershell_screenshot](https://user-images.githubusercontent.com/12364330/112062829-db611680-8b36-11eb-9778-262a2a563a0e.png)

### Task logs

[Worker_20210322-213505-utc.log](https://github.com/microsoft/azure-pipelines-tasks/files/6185416/Worker_20210322-213505-utc.log)
[Agent_20210322-213456-utc.log](https://github.com/microsoft/azure-pipelines-tasks/files/6185417/Agent_20210322-213456-utc.log)
[ReleaseLogs_7186.zip](https://github.com/microsoft/azure-pipelines-tasks/files/6185457/ReleaseLogs_7186.zip)

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

We stopped hearing from agent xxxxx. Verify the agent machine is running and has a healthy network connection. Anything that terminates an agent process, starves it for CPU, or blocks its network access can cause this error.

**Comment by andy-mishechkin** [comment]
Hello @begroff 
Have you got this error sometimes or always when you got a second task ?
I've checked log files, which you included to this ticket.
Would you please provide powershell script ToggleAppPool.ps1, which you tried to execute in task or perhaps we can fork your repository for investigation, if it possible ?
Also I see, you tried to stop\start IIS application pool in your task - what a kind of applications runs in this pool (for example - integrated mode standard ASP.NET application or classic for legacy web applications without managed code). May check, please, the correspond w3wp.exe process has any blockers ?
And I didn't see any failed tasks your worker log. May you provide worker log where you've got this error.

Thank you.

**Comment by andy-mishechkin** [comment]
Closed, because no any activity in 14 days.
@begroff , please feel free to reopen, if you'll have any updates

**Comment by begroff** [comment]
Andy,

Sorry for the late reply. We ended up solving our problem in a couple of ways. 

First after our update to the ADOS version mentioned in the ticket, we noticed that during builds our agent version was reverting back from version 2.184.1, to version 2.170.1 and multiple PowerShell tasks were failing. What we ended up doing is installing the agent version 2.170.1 directly and that seemed to fix the issue. We then realized that some of our PowerShell scripts were not using the $LASTEXITCODE variable at all and/or we were not exiting or exiting improperly.  Those scripts were updated and that seemed to fix the issue too.

I attached the worker logs that we received from our testing and peculiarly it didn't include any failures (even though the pipelines with the multiple PowerShell tasks reported failure). I won't be able to provide you with a worker log that has a failure. I am including our original PowerShell script that did not have exit codes, along with a new one that we created which does. We still need to make changes to the scripts regarding setting the $LASTEXITCODE variable when the script starts but having that without being set seems to have fixed the problem.

The application that we were trying to stop/start the app pool was a standard .NET framework web app in integrated mode.

[ToggleAppPool_Original.txt](https://github.com/microsoft/azure-pipelines-tasks/files/6387043/ToggleAppPool_Original.txt)
[ToggleAppPool.txt](https://github.com/microsoft/azure-pipelines-tasks/files/6387060/ToggleAppPool.txt)
